### PR TITLE
Radare2 - #1231

### PR DESCRIPTION
**Pull request by cldrn**
Added an intro to r2, installation instructions references for installing on systems and actual android devices, and basic commands for android analysis.

This PR covers issue #1231.

### CODE REVIEW

File: Document/0x05b-Basic-Security_Testing.md [hunk 1]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
 
 ##### radare2
 
--- ToDo: <https://github.com/OWASP/owasp-mstg/issues/1231>
+[radare2](https://rada.re/r/ "Radare2 official website") is a popular reverse engineering framework for analysis, disassembling, debugging and patching that is scriptable and supports many architectures and file formats including Android/iOS apps. For Android, Dalvik DEX (odex, multidex), ELF (executables, .so, ART) and Java classes are supported, and it even has a web server for collaborative work. It also contains several useful scripts that can help you during mobile application analysis as it offers low level disassembling and safe static analysis that comes handy when traditional tools fail.
+
+###### Installing radare2
+
+Please refer to [radare2's official installation instructions](https://github.com/radare/radare2/blob/master/README.md "radare2 installation instructions"). There are also installation instructions If you need to [install radare2 on Android](https://github.com/radare/radare2/blob/master/doc/android.md "Install radare2 on Android").
```
Comment (TheDauntless) [inline Document/0x05b-Basic-Security_Testing.md]: If -> if

File: Document/0x05b-Basic-Security_Testing.md [hunk 2]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
+
+###### Using radare2
+
+You can use `rafind2` to read strings directly from the encoded AndroidManifest.xml. Use this to list permissions, activities, providers, services, receivers and other information stored in AndroidManifest.xml
+
+```shell
```
Comment (TheDauntless) [inline Document/0x05b-Basic-Security_Testing.md]: Chang this to one large shell block where you add a # before each command to explain what you want to do:

```shell
# List all activities
$ rafind2 -ZS activity AndroidManifest.xml

# List all providers
$ rafind2 -ZS provider AndroidManifest.xml
```

File: Document/0x05b-Basic-Security_Testing.md [hunk 3]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
+[0x000009c8]> icq
+0x0000069c [0x0000078c - 0x000007b8] Lsg/vantagepoint/a/a Ljava/lang/Object;
+0x000006bc [0x000007c8 - 0x000007ec] Lsg/vantagepoint/a/b Ljava/lang/Object;
+...
+0x0000073c [0x00000958 - 0x00000abc] Lsg/vantagepoint/uncrackable1/MainActivity Landroid/app/Activity;
+0x0000075c [0x00000acc - 0x00000bb2] Lsg/vantagepoint/uncrackable1/a Ljava/lang/Object;
+```
+
+You can also print the external methods with the r2 command `iiq`.
```
Comment (TheDauntless) [inline Document/0x05b-Basic-Security_Testing.md]: remove 'also'

File: Document/0x05b-Basic-Security_Testing.md [hunk 4]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
+        ,=< 0x000009d0      39000e00       if-nez v0, 0x000009ec
+        |   0x000009d4      710028000000   invoke-static {}, Lsg/vantagepoint/a/c.b()Z
+        ...
+            ; 0x29
+       ||   0x000009e6      0a00           move-result v0
+      ,===< 0x000009e8      38000700       if-eqz v0, 0x000009f6
+      |``-> 0x000009ec      1a003f00       const-string v0, str.Root_detected ; 0x11bf
+```
+
+You can also print the class names with the r2 command `icq`.
```
Comment (TheDauntless) [inline Document/0x05b-Basic-Security_Testing.md]: Remove 'also'

File: Document/0x05b-Basic-Security_Testing.md [hunk 5]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
+[0x000009c8]> iiq
+Landroid/app/Activity.method.<init>()V
+Landroid/app/Activity.method.onCreate(Landroid/os/Bundle;)V
+...
+0x0000075c [0x00000acc - 0x00000bb2]    230 class 6 Lsg/vantagepoint/uncrackable1/a super: Ljava/lang/Object;
+0x00000acc method 0 sp   Lsg/vantagepoint/uncrackable1/a.method.a(Ljava/lang/String;)Z
+0x00000b5c method 1 sp   Lsg/vantagepoint/uncrackable1/a.method.b(Ljava/lang/String;)[B
+```
+
+You can also print all the strings with the r2 command `izq`.
```
Comment (TheDauntless) [inline Document/0x05b-Basic-Security_Testing.md]: Remove 'also'

File: Document/0x05b-Basic-Security_Testing.md [hunk 6]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
+0xd76 6 6 <init>
+0xd83 10 10 AES error:
+0xd8f 20 20 AES/ECB/PKCS7Padding
+0xda5 18 18 App is debuggable!
+0xdc0 9 9 CodeCheck
+0x11ac 7 7 Nope...
+0x11bf 14 14 Root detected!
+```
+
+Radare2 is very powerful and has dozens commands that you can find on the [radare2 command documentation](https://radare.gitbooks.io/radare2book/basic_commands/intro.html "radare2 command documentation").
```
Comment (TheDauntless) [inline Document/0x05b-Basic-Security_Testing.md]: dozens -> dozens of

File: Document/0x05b-Basic-Security_Testing.md (hunk 1 above)
Comment (cpholguera) [inline Document/0x05b-Basic-Security_Testing.md]: you could shortly mention that the installation on Android is done via Termux, as we talk about Termux in other places of this chapter. Nice to link things :)

File: Document/0x05b-Basic-Security_Testing.md (hunk 2 above)
Comment (cpholguera) [reply Document/0x05b-Basic-Security_Testing.md]: I agree, one for info. gathering with rafind2, one /more for r2 basics.

File: Document/0x05b-Basic-Security_Testing.md [hunk 7]
```diff
@@ -370,30 +370,18 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
-$ rafind2 -ZS provider AndroidManifest.xml
-```
-
-```shell
-$ rafind2 -ZS service AndroidManifest.xml
-```
-
-```shell
-$ rafind2 -ZS receiver AndroidManifest.xml
+$ rafind2 -ZS permission AndroidManifest.xml # Permissions
```
Comment (TheDauntless) [inline Document/0x05b-Basic-Security_Testing.md]: Put them before the command. Looks overkill now, but it's consistent :)

File: Document/0x05b-Basic-Security_Testing.md [hunk 8]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
+
+###### Installing radare2
+
+Please refer to [radare2's official installation instructions](https://github.com/radare/radare2/blob/master/README.md "radare2 installation instructions"). There are also installation instructions If you need to [install radare2 on Android](https://github.com/radare/radare2/blob/master/doc/android.md "Install radare2 on Android").
+
+###### Using radare2
+
+You can use `rafind2` to read strings directly from the encoded AndroidManifest.xml. Use this to list permissions, activities, providers, services, receivers and other information stored in AndroidManifest.xml
```
Comment (cpholguera) [inline Document/0x05b-Basic-Security_Testing.md]: you need to install axml2xml via r2pm, right?

File: Document/0x05b-Basic-Security_Testing.md (hunk 6 above)
Comment (cpholguera) [inline Document/0x05b-Basic-Security_Testing.md]: Not related specifically to this line:

Also mention that you can also use it for JNI

File: Document/0x05b-Basic-Security_Testing.md [hunk 9]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
+
+```shell
+$ rafind2 -ZS service AndroidManifest.xml
+```
+
+```shell
+$ rafind2 -ZS receiver AndroidManifest.xml
+```
+
+You can use the flag `-a` to manually set the architecture to Dalvik.
```
Comment (cpholguera) [inline Document/0x05b-Basic-Security_Testing.md]: Explain briefly here/above what's rafind2 and r2 and why you use it this way. Also you'd like to use rabin2 for some quick things and r2 to go deeper? We don't have to give much detail here. It should be explained properly in the Rev. Eng. chapter.

File: Document/0x05b-Basic-Security_Testing.md [hunk 10]
```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
...
+
+```shell
+$ r2 -a dalvik classes.dex
```
Comment (cpholguera) [inline Document/0x05b-Basic-Security_Testing.md]: r2 should recognize this automatically, right? I'd check it and remove the extra flag.

File: Document/0x05b-Basic-Security_Testing.md (hunk 8 above)
Comment (cldrn) [reply Document/0x05b-Basic-Security_Testing.md]: I didn't have to on an Ubuntu machine. Should I add this requirement?

File: Document/0x05b-Basic-Security_Testing.md (hunk 10 above)
Comment (cldrn) [reply Document/0x05b-Basic-Security_Testing.md]: It does. I added more documentation about how you only use this flag when detection does not works.

File: Document/0x05b-Basic-Security_Testing.md (hunk 9 above)
Comment (cldrn) [reply Document/0x05b-Basic-Security_Testing.md]: Done.

File: Document/0x05b-Basic-Security_Testing.md (hunk 6 above)
Comment (cldrn) [reply Document/0x05b-Basic-Security_Testing.md]: Added it to the beginning of the section.

File: Document/0x05b-Basic-Security_Testing.md (hunk 1 above)
Comment (cldrn) [reply Document/0x05b-Basic-Security_Testing.md]: Fixed.

File: Document/0x05b-Basic-Security_Testing.md (hunk 7 above)
Comment (cldrn) [reply Document/0x05b-Basic-Security_Testing.md]: Fixed.

**Pull request merged**